R/morphemepiece.R

```diff
@@ -126,67 +126,176 @@ magrittr::`%>%`
+#' word (controlled by the `dir` parameter). The input vocabulary must be split
+#' into prefixes, suffixes, and "words".
+#' 
+#' @param word Word to tokenize.
+#' @param vocab_split List of named integer vector containing vocabulary words.
+#'   Should have components named "prefixes", "words", "suffixes"
+#' @param dir Integer; if 1 (the default), look for tokens starting at the
+#'   beginning of the word. Otherwise, start at the end.
+#' @param unk_token Token to represent unknown words.
+#' @param max_chars Maximum length of word recognized.
 #'
+#' @return Input word as a list of tokens.
 #' @keywords internal
-.mp_tokenize_word <- function(word, vocab) {
-    # handle prefixes a little differently...
-    prefixes <- vocab[stringr::str_ends(vocab, "##")]
-    names(prefixes) <- stringr::str_sub(prefixes, end = -3)
+.mp_tokenize_word <- function(word, 
+                              vocab_split, 
+                              dir = 1, # -1 for backwards
+                              unk_token = "[UNK]",
+                              max_chars = 100) {
+  if (stringi::stri_length(word) > max_chars) {
+    return(unk_token)
+  }
+  frag_pat <- "##"
+  
+  prefixes <- vocab_split$prefixes
+  words <- vocab_split$words
+  suffixes <- vocab_split$suffixes
+  
+  is_bad  <- FALSE
+  start <- 1
+  sub_tokens <- character(0)
+  
+  wordlen <- stringi::stri_length(word)
+  end <- wordlen
+  
+  if (dir == 1) {
+    # rules for what kind of token can follow what (prefix, word, suffix)
+    allowed_next_rules <- list("p" = c("p", "w", "s"),
+                               "w" = "s",
+                               "s" = "s")
+    allowed_next <- c("p", "w")
+  } else {
+    # for backwards run
+    allowed_next_rules <- list("s" = c("p", "w", "s"),
+                               "w" = "p",
+                               "p" = "p")
     
-    # here, the names have the ##, because that's what's unique.
-    mp_vocab_nohash <- stringr::str_remove_all(vocab, "#")
-    names(mp_vocab_nohash) <- vocab
+    allowed_next <- c("s", "w")  
+  }
+  
+  keepgoing <- TRUE
+  while (keepgoing) {
+    if (dir == 1) {
+      end <- wordlen
+    } else {
+      start <- 1
+    }
     
-    current_breakdown <- word
-    keepgoing <- TRUE
-    while(keepgoing) {
-        current_breakdown <- current_breakdown[current_breakdown != ""]
-        piece_found <- current_breakdown %in% vocab
-        if (all(piece_found)) {
-            break
-        } else {
-            piece_to_work_on <- match(FALSE, piece_found)
-            current_word <- current_breakdown[piece_to_work_on]
-        }
-        original_start <- FALSE
-        if (piece_to_work_on == 1) {
-            original_start <- TRUE
-        }
-        original_end <- FALSE
-        if (piece_to_work_on == length(current_breakdown)) {
-            original_end <- TRUE
-        }
+    cur_substr  <- NA_character_
+    while (start <= end) {
```

Comment:
Nested `while` makes me think we can probably make this better. No worries for this check in but maybe add a ticket to dig into this part of the code for efficiency (here or in {wordpiece} as appropriate).

Reply:
I'll add a ticket, but I'll let you do the digging. :) I doubt that I could make this part much faster (whatever we did here would also apply to wordpiece).
https://github.com/macmillancontentscience/morphemepiece/issues/3